### PR TITLE
autopilot+lnrpc: add graph diameter calculation to getnetworkinfo

### LINKED ISSUE #180: Add graph diameter calculation to GetNetworkInfo
The response of the `GetNetworkInfo` RPC contains the following fields:
```proto
message NetworkInfo {
    uint32 graph_diameter = 1;
    double avg_out_degree = 2;
    uint32 max_out_degree = 3;
    uint32 num_nodes = 4;
    uint32 num_channels = 5;
    int64 total_network_capacity = 6;
    double avg_channel_size = 7;
    int64 min_channel_size = 8;
    int64 max_channel_size = 9;
}
```

Currently, all fields _except_ the `graph_diameter` field are populated. The [graph diameter](http://mathworld.wolfram.com/GraphDiameter.html) is to represent the diameter, or the worst-case path length from one node to another within the channel graph.

### PR DESCRIPTION
**Pull request by bitromortac**
Computes the Lightning Network's diameter using breadth-first-search (BFS) implemented in `autopilot.SimpleGraph` and via a more efficient method, and exposes it to the `getnetworkinfo` results. Fixes #180.

The calculation of the graph diameter (longest shortest path or maximal eccentricity in the graph) with all-pairs shortest path lengths is cumbersome as noted in previous work #417. It takes about 2 minutes to evaluate the whole graph on my machine, with some optimizations already in place. Other optimizations could be investigated, needing a method to generate and benchmark large graphs. We are looking for a ~100x improvement for the `getnetworkinfo` rpc call to return quickly, which requires probably more than only the optimization of BFS.

* Instead of using *all*-pairs shortest paths, we can cut out the interior of the graph, by measuring with respect to the highest-degree node (which is very central). The starting point is a single BFS from the central node, computing all distances to the other nodes. Then, by discarding nodes reachable within some percentage of the reference node's eccentricity, we only keep the outer rim of the graph (we could still change the rim definition). We then only compute shortest path lengths between all pairs of that subset, decreasing computation time greatly (to 100 ms - 1 s, depending on how large the outer node subset is).

* Another bottleneck is the initialization of a `SimpleGraph` from the channeldb (about 10 s). I added a new `autopilot.ChannelGraph` implementation that uses the cached channeldb data recently introduced. Strangely, cached and uncached graphs lead to different results of the diameter, 12 and 9, respectively. Is there different graph pruning applied?

## Alternatives:
* We could compute eccentricities of all nodes repeatedly in the background, adding them to the node metrics. The diameter can then be extracted as the maximal eccentricity.
* A more informative number would perhaps be the diameter of the innermost 95% of nodes.

## Logs:
Cached graph:
```
[INF] ATPL: initialized simple graph with 18215 nodes and 79681 channels
[INF] ATPL: cutoff radius is 6 hops (max-degree node's eccentricity is 8)
[INF] ATPL: evaluated nodes: 6, discarded nodes 18209
[INF] RPCS: elapsed time for cutoff-diameter (12) calculation: 55.042813ms
[INF] RPCS: elapsed time for exact diameter (12) calculation: 2m9.955505819s
```

Uncached graph:
```
[INF] ATPL: initialized simple graph with 17371 nodes and 74193 channels
[INF] ATPL: cutoff radius is 3 hops (max-degree node's eccentricity is 5)
[INF] ATPL: evaluated nodes: 143, discarded nodes 17228
[INF] RPCS: elapsed time for cutoff-diameter (9) calculation: 943.280863ms
[INF] RPCS: elapsed time for exact diameter (9) calculation: 1m32.118344346s
```

## Todo:
- [ ] figure out why cached and uncached graphs are different
- [ ] add tests for cached channeldb conversion?

#### Pull Request Checklist

- [x] All changes are Go version 1.16 compliant
- [ ] Your PR passes all CI checks. If a check cannot be passed for a justifiable reason, that reason must be stated in the commit message and PR description. (I was only able to run unit tests so far.)
- [x] If this is your first time contributing, we recommend you read the [Code Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md)
   - [x] The code being submitted is commented according to [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation)
   - [x] Commits have a logical structure according to [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure)
- [x] For new code: Code is accompanied by tests which exercise both the positive and negative (error paths) conditions (if applicable)
- [x] Any new logging statements use an appropriate subsystem and logging level
- [x] For code and documentation: lines are wrapped at 80 characters (the tab character should be counted as 8 characters, not 4, as some IDEs do per default)
- [ ] A description of your changes [should be added to running the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes) for the milestone your change will land in

### CODE REVIEW

**Comment by bhandras** [comment]
Thanks for the PR @bitromortac 

>  Strangely, cached and uncached graphs lead to different results of the diameter, 12 and 9, respectively. Is there different graph pruning applied?

Yes indeed, the cached graph is already heavily pruned.

**Review by bhandras — COMMENTED** [review]
Great work @bitromortac! :1st_place_medal: 
Overall PR is in a really good shape, added a few nits to fix before approval.

Please don't forget to update the `docs/release-notes/release-notes-0.15.0.md` file too.

File: autopilot/simple_graph.go [hunk 1]
```diff
@@ -62,5 +69,156 @@ func NewSimpleGraph(g ChannelGraph) (*SimpleGraph, error) {
...
+// degree determines the number of edges for a node in the graph.
+func (graph *SimpleGraph) degree(node int) int {
+	return len(graph.Adj[node])
+}
+
+// nodeMaxDegree determines the node with the max degree and its degree.
+func (graph *SimpleGraph) nodeMaxDegree() (int, int) {
+	var (degree, maxNode, maxDegree int)
+	for node, _ := range graph.Adj {
+		degree = graph.degree(node)
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply autopilot/simple_graph.go]: Done

File: autopilot/simple_graph.go [hunk 2]
```diff
@@ -62,5 +69,156 @@ func NewSimpleGraph(g ChannelGraph) (*SimpleGraph, error) {
...
+}
+
+// degree determines the number of edges for a node in the graph.
+func (graph *SimpleGraph) degree(node int) int {
+	return len(graph.Adj[node])
+}
+
+// nodeMaxDegree determines the node with the max degree and its degree.
+func (graph *SimpleGraph) nodeMaxDegree() (int, int) {
+	var (degree, maxNode, maxDegree int)
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply autopilot/simple_graph.go]: Done

File: rpcserver.go [hunk 3]
```diff
@@ -5822,10 +5822,20 @@ func (r *rpcServer) GetNetworkInfo(ctx context.Context,
...
+	// Graph diameter.
+	channelGraph := autopilot.ChannelGraphFromDatabase(graph)
+	simpleGraph, err := autopilot.NewSimpleGraph(channelGraph)
+	if err != nil {
+		return nil, err
+	}
+	start := time.Now()
+	diameter := simpleGraph.DiameterRadialCutoff()
+	rpcsLog.Infof("elapsed time for diameter (%d) calculation: %v", diameter,
+		time.Now().Sub(start))
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply rpcserver.go]: Changed it

File: autopilot/simple_graph.go [hunk 4]
```diff
@@ -1,5 +1,12 @@
 package autopilot
 
+// DIAMETER_CUTOFF is used to discard nodes in the diameter calculation.
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply autopilot/simple_graph.go]: Yes, that's very useful.

File: autopilot/simple_graph_test.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+	return SimpleGraph{Adj: adjList}
+}
+
+func TestShortestPathLengths(t *testing.T) {
+	simpleGraph := NewTestSimpleGraph(centralityTestGraph)
+
+	// Test the shortest path lengths from node 0 to all other nodes.
+	shortestPathLengths := simpleGraph.shortestPathLengths(0)
+	eq := reflect.DeepEqual(shortestPathLengths, testShortestPathLengths)
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply autopilot/simple_graph_test.go]: Thank you, good to know

**Comment by bitromortac** [comment]
Thank you, @bhandras, for the very useful hints!

An alternative to the 75% cutoff could be to take, say the 500 outermost nodes (with respect to the central node) into account. That would make the command run in about constant time, instead of being somewhat unpredictable for how many nodes the eccentricity is calculated.

**Review by bhandras — APPROVED** [review]
LGTM :1st_place_medal:

File: autopilot/simple_graph.go [hunk 6]
```diff
@@ -62,5 +69,156 @@ func NewSimpleGraph(g ChannelGraph) (*SimpleGraph, error) {
...
 	}
 
+	// We prepare to give some debug output about the size of the graph.
+	totalChannels := 0
+	for _, channels := range graph.Adj {
+		totalChannels += len(channels)
+	}
+
+	// The number of channels is double counted, so divide by two.
+	log.Debugf("initialized simple graph with %d nodes and %d channels",
```
Comment (guggero) [inline autopilot/simple_graph.go:79]: nit: log statements (in contrast to error messages) usually start capitalized.

File: autopilot/graph.go [hunk 7]
```diff
@@ -455,6 +455,97 @@ func (m *memChannelGraph) addRandNode() (*btcec.PublicKey, error) {
...
+	return &databaseChannelGraphCached{
+		db: db,
+	}
+}
+
+// dbNodeCached is a wrapper struct around a database transaction a
+// channeldb.LightningNode. The wrapper methods implement the autopilot.Node
+// interface.
+type dbNodeCached struct {
+	node     *route.Vertex
```
Comment (guggero) [inline autopilot/graph.go:480]: This can be `route.Vertex` directly, no need for a pointer.

**Review by guggero — COMMENTED** [review]
Sorry for the delay on the review!
Code looks pretty good! I'm a bit concerned about this being in the `autopilot` package when we only currently use the new functionality for a single value in the network statistics RPC. But I can see why it was done that way to re-use a lot of existing code. So I guess I'm fine with it.

File: autopilot/simple_graph.go (hunk 6 above)
Comment (bitromortac) [reply autopilot/simple_graph.go]: Changed it to capital letters.

File: autopilot/simple_graph.go [hunk 8]
```diff
@@ -62,5 +69,156 @@ func NewSimpleGraph(g ChannelGraph) (*SimpleGraph, error) {
...
+	return maxNode, maxDegree
+}
+
+// shortestPathLengths performs a breadth-first-search from a node to all other
+// nodes, returning the lengths of the paths.
+func (graph *SimpleGraph) shortestPathLengths(node int) map[int]uint32 {
+	// level indicates the shell of the search around the root node.
+	var level uint32
+	graphOrder := len(graph.Adj)
+	// nextLevel tracks which nodes should be visited in the next round.
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply autopilot/simple_graph.go:120]: Added the newlines. I didn't spot the character limit violations, could you point to an example?

File: autopilot/simple_graph.go [hunk 9]
```diff
@@ -62,5 +69,156 @@ func NewSimpleGraph(g ChannelGraph) (*SimpleGraph, error) {
...
+	// have more nodes to explore.
+	for len(nextLevel) > 0 {
+		level++
+
+		// We swap the queues for efficient memory management.
+		thisLevel, nextLevel = nextLevel, thisLevel
+
+		// Visit all neighboring nodes of the level and mark them as seen if
+		// they were not discovered before.
+		for _, thisNode := range thisLevel {
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply autopilot/simple_graph.go:145]: I determined shortest paths from the test graph manually, I think it should be OK.

File: autopilot/graph.go (hunk 7 above)
Comment (bitromortac) [reply autopilot/graph.go]: Done

File: autopilot/simple_graph.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (guggero) [reply autopilot/simple_graph.go:120]: Line 143 ("Visit all neighboring nodes...") is an example. I know this is very very nit-picky. It's just usually a sign that the editor wasn't set up properly to use 8 spaces for the tab character (and show a caret line at 80 characters, which makes life so much easier).

[the remark this answers is not in the document]
Comment (bitromortac) [reply autopilot/simple_graph.go:120]: Not picky at all, thank you. I just re-read the guidelines today (before seeing your comment) and spotted the 8 character tab size, indeed I had set it to 4, will correct it.

**Review by guggero — APPROVED** [review]
LGTM :tada:

**Pull request merged**